### PR TITLE
proactive TWAP trap example

### PR DESCRIPTION
**Pull request by An0nJon**
Trap example to demonstrate a proactive detection of a multi-block oracle manipulation using time-series data.

### CODE REVIEW

File: oracles/twap/src/TWAPTrap.sol [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import {Test, console2} from "forge-std/Test.sol";
```
Comment (scglenn) [inline oracles/twap/src/TWAPTrap.sol:4]: This isnt needed in the Trap.sol right?

File: oracles/twap/src/TWAPTrap.sol [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * @dev Collects the latest price data point.
+     * @return The latest price data point.
+     * @dev This function is a placeholder and returns a dummy value.
+     * In a real implementation, it would fetch the latest price data from an oracle.
+     */
+    function collect() external view returns (PriceDataPoint memory) {
+        return PriceDataPoint({price: 0});
```
Comment (scglenn) [inline oracles/twap/src/TWAPTrap.sol:37]: I would prefer to query a mockOracle so we avoid confusion with some of the early examples of traps

**Review by scglenn — CHANGES_REQUESTED** [review]

File: oracles/twap/src/TWAPTrap.sol [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * @dev Checks if the provided price data points are valid based on the deviation threshold.
+     * @param dataPoints An array of price data points to check.
+     * @return A boolean indicating whether the price data points are valid.
+     */
+    function isValid(PriceDataPoint[] calldata dataPoints) external view returns (bool) {
+        uint256 dataPointsLength = dataPoints.length;
+        if (dataPointsLength < 2 || dataPointsLength < dataPointsToCheck) {
```
[the remark this answers is not in the document]
Comment (An0nJon) [reply oracles/twap/src/TWAPTrap.sol:47]: I see. Ill rip that part out.

File: oracles/twap/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+## TWAP Oracle Trap
+
+To detect and prevent such price manipulation attacks, we can utilize Drosera's trap concept and create a trap to do time-series analysis on the price data from the TWAP oracle. The trap will continuously monitor the price data and compare it against predefined thresholds to identify any sudden and significant price deviations. If a deviation is detected, the trap will raise an alert, indicating a potential price manipulation attempt.
+
+We have implemented a proof-of-concept TWAP oracle trap in Solidity, and a simulated test contract to demonstrate its effectiveness in detecting the price manipulation that occurred during this potential attack.
+
+1. The `TWAPTrap.sol` contract:
+
+- This contract interacts with a hypothetical oracle to collect price data for each block.
+- It provides a `collect` function that is not used to retrieve mock oracle data.
```
Comment (scglenn) [inline oracles/twap/README.md:28]: the collect function is used instead of a mock now, remove this line

**Review by scglenn — CHANGES_REQUESTED** [review]

**Review by scglenn — APPROVED** [review]

**Pull request merged**